internal/network/netns.go

```diff
@@ -21,94 +10,100 @@ import (
+	if err := netlink.LinkSetNsPid(link, pid); err != nil {
+		return errors.Wrapf(err, "netlink.SetNsPid(%#v, %d) failed", link, pid)
 	}
+	return nil
+}
 
-	// Lock the OS Thread so we don't accidentally switch namespaces
+// DoInNetNS is a utility to run a function `work` inside of a specific network namespace
+// `ns`. This is accomplished by locking the current goroutines thread to prevent the goroutine
+// from being scheduled to a new thread during execution of `work`. The threads original network namespace
+// will be rejoined on exit.
+func DoInNetNS(ns netns.NsHandle, run func() error) error {
 	runtime.LockOSThread()
 	defer runtime.UnlockOSThread()
 
-	// Stash current network namespace away and make sure we enter it as we leave
-	log.Infof("Obtaining current namespace")
-	origNS, err := netns.Get()
+	origNs, err := netns.Get()
 	if err != nil {
-		return fmt.Errorf("netns.Get() failed: %v", err)
+		return errors.Wrap(err, "failed to get current network namespace")
 	}
-	defer origNS.Close()
-	log.Infof("Original namespace %v", origNS)
+	defer origNs.Close()
 
-	// Get a reference to the new network namespace
-	ns, err := netns.GetFromPid(*nspid)
-	if err != nil {
-		return fmt.Errorf("netns.GetFromPid(%d) failed: %v", *nspid, err)
+	if err := netns.Set(ns); err != nil {
+		return errors.Wrapf(err, "failed to set network namespace to %v", ns)
 	}
-	defer ns.Close()
-	log.Infof("New network namespace from PID %d is %v", *nspid, ns)
+	// Defer so we can re-enter the threads original netns on exit.
+	defer netns.Set(origNs)
 
-	// Get a reference to the interface and make sure it's down
-	link, err := netlink.LinkByName(*ifStr)
-	if err != nil {
-		return fmt.Errorf("netlink.LinkByName(%s) failed: %v", *ifStr, err)
+	return run()
+}
+
+// NetNSConfig moves a network interface into a network namespace and
+// configures it. The configuration specified by `adapterCfg` is expected to be
+// passed in as a JSON object (marshalled prot.NetworkAdapter). This function will log
+// it's output to `logFile`.
+//
+// This function MUST be used in tandem with `DoInNetNS` or some other means that ensures that the goroutine
+// executing this code stays on the same thread.
+func NetNSConfig(ifStr string, nsPid int, adapterCfg string, logFile string) error {
+	if ifStr == "" || nsPid == -1 || adapterCfg == "" || logFile == "" {
+		return errors.New("All four arguments must be specified")
 	}
-	if err := netlink.LinkSetDown(link); err != nil {
-		return fmt.Errorf("netlink.LinkSetDown(%#v) failed: %v", link, err)
+
+	logFileHandle, err := os.OpenFile(logFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return errors.Wrap(err, "failed to create log file")
 	}
+	defer logFileHandle.Close()
 
-	// Move the interface to the new network namespace
-	if err := netlink.LinkSetNsPid(link, *nspid); err != nil {
-		return fmt.Errorf("netlink.SetNsPid(%#v, %d) failed: %v", link, *nspid, err)
+	log := logrus.New()
+	log.SetOutput(logFileHandle)
```

Comment:
Do we need this now that we don't have to run a separate executable in the UVM? There's nothing that reads this file afaik anyways, I assume it's original intention was for debugging on our side.

Reply:
I wanted to keep it for debuggability but if no one cares or has ever found this useful I can just log as normal

Reply:
It's helped me in a few cases (but that's because I was specifically making changes TO this codepath 😆)